From f581917c07157230cb32e48333d12418355d6b63 Mon Sep 17 00:00:00 2001
From: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Date: Wed, 24 Oct 2012 17:33:32 +0200
Subject: [PATCH 41/78] wifi: balance calls to list_add/list_del if interface
 add fails

BZ: 63774

if add_interface fails, mac80211 layer will call drv_stop. In that
case, we are not executing the balanced list_del

Change-Id: Ia8b8e7efe1f6b0b328c4ff1b36e309a8dbd5f4eb
Signed-off-by: Jonathan DE CESCO <jonathan.de.cesco@intel.com>
Reviewed-on: http://android.intel.com:8080/71523
Reviewed-by: Casasnovas, QuentinX <quentinx.casasnovas@intel.com>
Reviewed-by: Regairaz, Julie <julie.regairaz@intel.com>
Tested-by: Champciaux, Nicolas <nicolas.champciaux@intel.com>
Reviewed-by: buildbot <buildbot@intel.com>
Tested-by: buildbot <buildbot@intel.com>
Reviewed-on: http://android.intel.com:8080/77998
Reviewed-by: Li, XiaojingX <xiaojingx.li@intel.com>
Tested-by: Li, XiaojingX <xiaojingx.li@intel.com>
---
 wl12xx-compat/drivers/net/wireless/wl12xx/main.c |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
index c1d5aec..fdd9139 100644
--- a/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
+++ b/wl12xx-compat/drivers/net/wireless/wl12xx/main.c
@@ -2404,7 +2404,16 @@ static void wl1271_op_stop_locked(struct wl1271 *wl)
 		if (test_and_clear_bit(WL1271_FLAG_RECOVERY_IN_PROGRESS,
 				       &wl->flags))
 			wl1271_enable_interrupts(wl);
-
+		else {
+                       /*
+                       * Delete wl from global wl_list since it has been
+                       * added in op_start, even if add interface failed
+                       */
+                       wl1271_debug(DEBUG_MAC80211, "mac80211 stop locked in state off");
+                       mutex_lock(&wl_list_mutex);
+                       list_del(&wl->list);
+                       mutex_unlock(&wl_list_mutex);
+		}
 		return;
 	}
 
-- 
1.7.9

